training/models/training_action.py

```diff
@@ -0,0 +1,151 @@
+    action_ids = fields.One2many(
+        M % "action",
+        "type_id",
+        "Training actions",
+        help="Training actions of this type.")
+    expected_duration_type_ids = fields.Many2many(
+        M % "duration_type",
+        string="Expected hour types",
+        help="These types of hours are expected in this type of training "
+             "action. For example, a training of type 'mixed' may expect "
+             "hours of types 'on-site' and 'online'.")
+
+
+class Action(models.Model):
+    """Define training actions.
+
+    A training action is one course that your company teaches and has in its
+    catalog of courses.
+
+    They define some requirements that the corresponding event is expected to
+    fulfill. Events linked to training actions are considered training groups.
+    """
+    _name = M % "action"
+
+    name = fields.Char(required=True, index=True, translate=True)
+    type_id = fields.Many2one(
+        M % "action_type",
+        "Training type",
+        required=True)
+    contents = fields.Html(
+        translatable=True,
+        help="Contents of the course, shown in the rear side of the diploma.")
+    append_template = fields.Selection(
+        ((2, "2 columns"), (3, "3 columns"), (4, "4 columns")),
+        help="Append one of these templates to the diploma contents. "
+             "They will help you to achieve some complex layouts.")
+    duration_ids = fields.One2many(
+        M % "duration",
+        "action_id",
+        "Expected hours",
+        copy=True,
+        help="Expected duration of each type of hours for these training "
+             "groups.")
+    event_ids = fields.One2many("event.event", "training_action_id", "Events")
+    grade_min = fields.Float(
+        "Minimal grade",
+        default=0,
+        required=True,
+        help="Students cannot get less than this grade.")
+    grade_pass = fields.Float(
+        "Passing grade",
+        default=5,
+        required=True,
+        help="Students above than this grade will pass the training.")
+    grade_max = fields.Float(
+        "Maximum grade",
+        default=10,
+        required=True,
+        help="Students cannot get more than this grade.")
+    material_ids = fields.Many2many(
+        M % "material",
+        string="Materials",
+        help="These should be delivered to every student in this training.")
+
+    @api.constrains("grade_min", "grade_pass", "grade_max")
+    def _check_grade_limits(self):
+        """Ensure no conflicts with grade limits."""
+        for s in self:
+            # Grade limits must be coherent
+            if not s.grade_min <= s.grade_pass <= s.grade_max:
+                msg = (_("Minimum grade cannot be bigger than maximum.")
+                       if s.grade_min > s.grade_max
+                       else _("Passing grade must be between minimum and "
+                              "maximum grades."))
+                raise ex.GradeLimitIncoherentError(msg)
```

Comment:
If you check the tests, you will see that they use all different exceptions in this module to check they are raised when needed. Without concrete exceptions, any other ValidationError would pass the test.